src/transformers/modeling_utils.py

```diff
@@ -1204,6 +1215,74 @@ def _from_config(cls, config, **kwargs):
 
         return model
 
+    def use_attn_implementation(self, attn_implementation: str):
+        """
+        Specifies the attention implementation to use in the model.
+
+        Args:
+            attn_implementation (`str`):
+                The attention implementation to use. Can be any of "eager" (manual implementation of the attention), "sdpa" (attention using [`torch.nn.functional.scaled_dot_product_attention`](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention.html)), or "flash_attention_2" (attention using [Dao-AILab/flash-attention](https://github.com/Dao-AILab/flash-attention)).
+        """
+        # TODO: Implement it. An implementation could be to define `self._eager_attn_class = XXXAttention`, `self._sdpa_attn_class = XXXSdpaAttention`, `self._flash_attn_class = XXXFlashAttention2` in the __init__ of XXXPreTrainedModel, and leverage those attributes here to replace the correct submodules. Alternatively, the `XXX_ATTENTION_CLASSES` in each modeling file can be leveraged.
+        raise NotImplementedError("model.use_attn_implementation is currently not implemented.")
+
+        if attn_implementation == "sdpa":
+            self.config = self._check_and_enable_sdpa(self.config, hard_check_only=False)
+        elif attn_implementation == "flash_attention_2":
+            # TODO: define torch_dtype properly
+            torch_dtype = None
+            self.config = self._check_and_enable_flash_attn_2(
+                self.config,
+                torch_dtype=torch_dtype,
+                device_map=getattr(self, "hf_device_map", None),
+                hard_check_only=False,
+            )
+
+    @classmethod
+    def _autoset_attn_implementation(
+        cls,
+        config,
+        use_flash_attention_2: Optional[bool] = None,
+        torch_dtype: Optional[torch.dtype] = None,
+        device_map: Optional[Union[str, Dict[str, int]]] = None,
+        check_device_map: bool = True,
+    ):
+        """
+        Automatically checks and dispatches to a default attention implementation. In order of priority:
+            1. An implementation specified in `config.attn_implementation`.
+            2. If specified, flash attention through use_flash_attention_2=True.
+            3. SDPA implementation, if available and supported by the model type.
+            4. Manual implementation otherwise.
+        """
+        # Here we use config._attn_implementation to check whether the attention implementation was explicitely set by the user.
+        # The property `PretrainedConfig.attn_implementation` is never `None`, for backward compatibility.
+        if hasattr(config, "_attn_implementation") and config._attn_implementation is not None:
```

Comment:
the config cannot no have this attribute anymore no?

Reply:
See: https://github.com/huggingface/transformers/pull/26572/files#r1412207800

Ok for me to do in follow-up PR